text/0000-conservative-variadics.md

```diff
@@ -0,0 +1,228 @@
+either because they were too broad, or had performance penalties that seemed out
+of place for a systems language.
+
+[#376]: https://github.com/rust-lang/rfcs/issues/376
+[#960]: https://github.com/rust-lang/rfcs/issues/960
+
+Without language level support for variadic functions, the closest alternative
+is to pass a slice. There are two major drawbacks to this approach:
+
+- If you want to take ownership of the values passed, you will have to move it
+  to the heap, or require that the type be `Copy`
+- The values must be of the same type. If you wish to take values which conform
+  to a trait, you must pass a trait object which at minimum means dynamic
+  dispatch, and could potentially mean boxing. Object safety also tends to be
+  pervasive throughout a library.
+
+To provide as conservative a feature as possible, this RFC will focus purely on
+the case where the types differ. Taking ownership of a variable number of values
+of the same type where that type is not `Copy` is a much smaller use case, and
+one that will likely be resolved by [RFC #1909], [RFC #1915], or some other RFC
+which allows `[T]` to be passed directly.
+
+[RFC #1909]: https://github.com/rust-lang/rfcs/pull/1909
+[RFC #1915]: https://github.com/rust-lang/rfcs/pull/1915
+
+We will look at two concrete library cases which would be reflected by this RFC.
+
+The first is the signature of [`Statement#execute` from the rust-postgres
+crate](https://docs.rs/postgres/0.13.4/postgres/stmt/struct.Statement.html#method.execute).
+This function takes `&[&ToSql]` as its final parameter, as an attempt to emulate
+variadic functions but avoid allocation. This leads to an API which is often
+difficult to use, and forces dynamic dispatch.
+
+The second case we will look at is a [proposal by the `diesel`
+crate](https://github.com/diesel-rs/diesel/pull/747) which is in a direction
+similar to this RFC. Their API focuses on using tuples, which they found painful
+for a variety of reasons.
+
+Ultimately both of these crates are trying to emulate a variadic function, and
+are creating APIs which are more painful to use as a result.
+
+# Detailed design
+[design]: #detailed-design
+
+We need to answer two questions to provide variadic functions. What does
+declaring a function as variadic look like, and how do we represent it?
+
+Variadic arguments would be represented as a heterogeneous list, often known as
+an hlist, which we will call a "variadic list" for this feature. The concrete
+types would live in `core::ops::variadic_list`, and would be defined as:
+
+```rust
+pub struct Cons<T, U>(pub T, pub U);
+pub struct Nil;
+```
+
+This RFC has chosen to introduce a new type for this, rather than attempting to
+use tuples as has been proposed in the past (notably by [RFC #1582]). Ultimately
+any usage of tuples for this would require treating them as an hlist to be at
+all ergonomic. Rather than introducing additional magic around tuples, it makes
```

Comment:
@sgrif That's true! When I said "fairly similar" I was referring to the ease/readability of the implementation. But you're correct, coherence does cause a number of problems in these cases (I've been stumbling on [this one](https://github.com/rust-lang/rust/issues/19032) recently myself).  

I'd prefer to resolve those coherence issues rather than designing new language features to avoid them. However, your design for variadics would provide a more immediate solution for those cases, which is a plus.